web/app/common/components/log-viewer/log-line/log-line.component.ts

```diff
@@ -0,0 +1,78 @@
+import { Component, AfterViewInit, ViewChild, ElementRef, Input, ViewEncapsulation } from '@angular/core';
+
+const ANSI_PATTERN = /\u001b\[([0-9;]+)?m/;
+
+export interface LogLine {
+  timestamp: number;
+  message: string;
+  level: string;
+  status: number;
+}
+
+@Component({
+  selector: 'fci-log-line',
+  templateUrl: './log-line.component.html',
+  styleUrls: ['./log-line.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+
+export class LogLineComponent implements AfterViewInit {
+  @Input() log: LogLine;
+  @ViewChild('logLine', { read: ElementRef }) logLineEl: ElementRef;
+
+  ngAfterViewInit() {
+    const parentEl = this.logLineEl.nativeElement;
+    let currentSpanEl = parentEl;
+    const stack = this.tokenize(this.log.message);
+
+    stack.forEach(tuple => {
+      const [code, text] = tuple;
+      if (code === '0') {
+        parentEl.innerHTML += text;
+        currentSpanEl = parentEl;
+      }
+      else {
+        currentSpanEl = this.injectSpan(currentSpanEl, text, code);
+      }
+    });
+  }
+
+  /**
+   * tokenization works by splitting the string by a regex that has a capture group
+   * this will return an flat array of tuples (string, capture)
+   * this method will split and group the result into an array of tuples [ansi code, text]
+   * NOTE: that the capture comes after the string, so we transpose them.
+   **/
+  private tokenize(text: string): [string, string][] {
+    const tuples = text.split(ANSI_PATTERN);
+
+    // if the text starts with a match capture (and thus returning empty string as [0]),
+    // use that captured ansi code as the beginning style
+    if (tuples[0] === '') {
+      tuples.shift();
+    }
+    // otherwise, we must assume we are starting each line as the default '0'
+    else {
+      tuples.unshift('0');
+    }
+
+    const stack = [];
+
+    for (let i = 0; i < tuples.length; i += 2) {
+      const code = tuples[i];
+      const text = tuples[i + 1];
+      stack.push([code, text]);
+    }
+    return stack;
+  }
+
+  private injectSpan(parent: HTMLElement, text: string, ansiCode: string): HTMLSpanElement {
+    const span = document.createElement('span');
```

Comment:
can you provide the document in the constructor. We should be using injectables instead of the direct global var

ex. https://github.com/fastlane/ci/blob/master/web/app/onboard/onboard.component.ts#L29